sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,481 @@
+            {"value": tree_id},
+        )
+        return r.rowcount
+
+    def _write_meta(self, sess, dataset):
+        """Write the title. Other metadata is not stored in a SQL Server WC."""
+        self._write_meta_title(sess, dataset)
+
+    def _write_meta_title(self, sess, dataset):
+        """Write the dataset title as a comment on the table."""
+        # TODO - dataset title is not stored anywhere in SQL server working copy right now.
+        # We can probably store it using function sp_addextendedproperty to add property 'MS_Description'
+        pass
+
+    def delete_meta(self, dataset):
+        """Delete any metadata that is only needed by this dataset."""
+        # There is no metadata stored anywhere except the table itself.
+        pass
+
+    def _get_geom_extent(self, sess, dataset, default=None):
+        """Returns the envelope around the entire dataset as (min_x, min_y, max_x, max_y)."""
+        geom_col = dataset.geom_column_name
+        r = sess.execute(
+            f"""
+            WITH _E AS (
+                SELECT geometry::EnvelopeAggregate({self.quote(geom_col)}) AS envelope
+                FROM {self.table_identifier(dataset)}
+            )
+            SELECT
+                envelope.STPointN(1).STX AS min_x,
+                envelope.STPointN(1).STY AS min_y,
+                envelope.STPointN(3).STX AS max_x,
+                envelope.STPointN(3).STY AS max_y
+            FROM _E;
+            """
+        )
+        result = r.fetchone()
+        if result == (None, None, None, None) and default is not None:
+            return default
+        return result
+
+    def _grow_rectangle(self, rectangle, scale_factor):
+        # scale_factor = 1 -> no change, >1 -> grow, <1 -> shrink.
+        min_x, min_y, max_x, max_y = rectangle
+        centre_x, centre_y = (min_x + max_x) / 2, (min_y + max_y) / 2
+        min_x = (min_x - centre_x) * scale_factor + centre_x
+        min_y = (min_y - centre_y) * scale_factor + centre_y
+        max_x = (max_x - centre_x) * scale_factor + centre_x
+        max_y = (max_y - centre_y) * scale_factor + centre_y
+        return min_x, min_y, max_x, max_y
+
+    def _create_spatial_index_post(self, sess, dataset):
+        # Only implementing _create_spatial_index_post:
+        # We need to know the rough extent of the data to create an index in that area.
+
+        L = logging.getLogger(f"{self.__class__.__qualname__}._create_spatial_index")
+
+        geom_col = dataset.geom_column_name
+        index_name = f"{dataset.table_name}_idx_{geom_col}"
+
+        # Create the SQL Server Spatial Index
+        L.debug("Creating spatial index for %s.%s", dataset.table_name, geom_col)
+        t0 = time.monotonic()
+        # Create an index over the extent of existing geometries + 20% room to grow.
+        GROW_FACTOR = 1.2
+        min_x, min_y, max_x, max_y = self._grow_rectangle(
+            self._get_geom_extent(sess, dataset, default=(-180, -90, +180, +90)),
```

Comment:
Unless the CRS is geographic the default here won't be useful... would be better _not_ to create a spatial index for an empty table with a geometry column.

Reply:
Okay, done